app/scripts/lib/WeakRefObjectMap.ts

```diff
@@ -0,0 +1,218 @@
+ * unavailable (i.e., garbage collected) independently of one another. Users of this map should be prepared to handle
+ * cases where a property's value has been collected and is therefore `undefined`.
+ *
+ * This class was implemented to help with memory management of network client proxies used by the SelectedNetworkController
+ * to keep per domain selected networks in sync. The properties of the NetworkClient object (provider and blockTracker) are weakly
+ * referenced so that they can be garbage collected if/when a dapp connection ends without effective cleanup.
+ */
+
+export class WeakRefObjectMap<RecordType extends Record<string, object>>
+  implements Map<string, RecordType>
+{
+  /**
+   * Internal map to store keys and their corresponding weakly referenced object values.
+   */
+  private map: Map<string, WeakRefObject<RecordType>>;
+
+  constructor() {
+    this.map = new Map();
+  }
+
+  /**
+   * Associates a key with a value in the map. If the key already exists, its associated value is updated.
+   * The values are stored as weak references.
+   *
+   * @param key - The key under which to store the value.
+   * @param value - The value to store under the specified key. Must be an object.
+   * @returns The `WeakRefObjectMap` instance.
+   */
+  set(key: string, value: RecordType): this {
+    const weakRefValue: Partial<WeakRefObject<RecordType>> = {};
+    for (const keyValue in value) {
+      if (!Object.hasOwn(value, keyValue)) {
+        continue;
+      }
+      const item: RecordType[typeof keyValue] = value[keyValue];
+      if (typeof item === 'object' && item !== null) {
+        weakRefValue[keyValue] = new WeakRef(item);
+      } else {
+        throw new Error(
+          `Property ${String(
+            keyValue,
+          )} is not an object and cannot be weakly referenced.`,
+        );
+      }
+    }
+    this.map.set(key, weakRefValue as WeakRefObject<RecordType>);
+    return this;
+  }
+
+  /**
+   * Retrieves the value associated with the specified key. The value is dereferenced before being returned.
+   * If the key does not exist or the value has been garbage collected, `undefined` is returned.
+   *
+   * @param key - The key whose associated value is to be returned.
+   * @returns The dereferenced value associated with the key, or `undefined`.
+   */
+  get(key: string): RecordType | undefined {
+    const weakRefValue = this.map.get(key);
+    if (!weakRefValue) {
+      return undefined;
+    }
+
+    const deRefValue: RecordType = {} as RecordType;
+    for (const keyValue in weakRefValue) {
+      if (!Object.hasOwn(weakRefValue, keyValue)) {
+        continue;
+      }
+      const deref = weakRefValue[keyValue].deref();
+      if (deref === undefined) {
+        this.map.delete(key);
+        return undefined;
+      }
+      deRefValue[keyValue] = deref;
+    }
+
+    return deRefValue;
```

Comment:
Nit: One more case where we can improve type safety by moving the assertion to after we've finished assembling the object, using `Partial` in the interim
```suggestion
    const deRefValue: Partial<RecordType> = {};
    for (const keyValue in weakRefValue) {
      if (!Object.hasOwn(weakRefValue, keyValue)) {
        continue;
      }
      const deref = weakRefValue[keyValue].deref();
      if (deref === undefined) {
        this.map.delete(key);
        return undefined;
      }
      deRefValue[keyValue] = deref;
    }

    return deRefValue as RecordType;
```